Count leading zero bits in one CPU tick

hnarkaytis opened the pull request:
Count of leading bits was implemented as a CPU instructing 6 years ago. It doesn't make sense to have custom code for this.
https://en.wikipedia.org/wiki/Bit_Manipulation_Instruction_Sets#ABM_(Advanced_Bit_Manipulation)

hnarkaytis committed:
Count leading zero bits in one CPU tick

hnarkaytis committed:
Count leading zero bits in one CPU tick

hnarkaytis committed:
Count leading zero bits in one CPU tick

remilapeyre reviewed (commented):
Is this supported on all architecture that Python support (https://www.python.org/download/other/)?
What about CPU that don't have this instruction or where this instruction is slower?
Also this would need to be compatible with all compilers supported by CPython.

hnarkaytis committed:
Merge branch 'count_leading_zeros' of github.com:hnarkaytis/cpython into count_leading_zeros

hnarkaytis committed:
Add support for _MSC_

hnarkaytis committed:
Workaround _MCVC_ support

mdickinson commented on Python/pymath.c:
```diff
@@ -80,17 +80,13 @@ round(double x)
-   }
-   d_bits += BitLengthTable[d];
-   return d_bits;
+   return d ? CHAR_BIT * sizeof (d) - __builtin_clz (d) : 0;
```
This is the wrong function. You want `__builtin_clzl`, not `__builtin_clz`. But in any case, this isn't going to work unless there's also a fallback for platforms that don't provide `__builtin_clz`.

aeros committed:
bpo-34037: Add Python API whatsnew for loop.shutdown_default_executor() (#19634)

Co-Authored-By: Victor Stinner <vstinner@python.org>

rhettinger committed:
Small improvements to the recipes and examples. (GH-19635)

* Add underscores to long numbers to improve readability
* Use bigger dataset in the bootstrapping example
* Convert single-server queue example to more useful multi-server queue

rhettinger committed:
bpo-40327: Improve atomicity, speed, and memory efficiency of the items() loop (GH-19628)

ned-deily committed:
bpo-40164: Update macOS installer builds to use OpenSSL 1.1.1g. (GH-19642)

hnarkaytis committed:
Workaround _MCVC_ inability to handle macro

hroncok committed:
bpo-38439: Add 256px IDLE icon (GH-17473)

Icon author: Andrew Clover, bpo-1490384

jmroot committed:
bpo-38360: macOS: support alternate form of -isysroot flag (GH-16480)

It is possible to use either '-isysroot /some/path' (with a space) or
'-isysroot/some/path' (no space in between). Support both forms in
places where special handling of -isysroot is done, rather than just
the first form.
Co-authored-by: Ned Deily <nad@python.org>

ned-deily committed:
bpo-38329: python.org macOS installers now update Current symlink (GH-19650)

Previously, python.org macOS installers did not alter the Current version
symlink in /Library/Frameworks/Python.framework/Versions when installing
a version of Python 3.x, only when installing 2.x.  Now that Python 2 is
retired, it's time to change that.  This should make it a bit easier
to embed Python 3 into other macOS applications.

hnarkaytis committed:
Count leading zero bits with single CPU instruction

hnarkaytis committed:
Replace __builtin_clz with __builtin_clzl

hnarkaytis committed:
Change preprocessor check on ifdef MS_WINDOWS

hnarkaytis committed:
Add debug preprocessor check on ifdef MS_WINDOWS

hnarkaytis committed:
Fix typo and argument type

vstinner committed:
bpo-40138: Fix Windows os.waitpid() for large exit code (GH-19637)

Fix the Windows implementation of os.waitpid() for exit code
larger than "INT_MAX >> 8". The exit status is now interpreted as an
unsigned number.

os.waitstatus_to_exitcode() now accepts wait status larger than
INT_MAX.

zooba committed:
bpo-40214: Fix ctypes WinDLL test with insecure flags (GH-19652)

isidentical committed:
bpo-39562: Prevent collision of future and compiler flags (GH-19230)

The constant values of future flags in the __future__ module
is updated in order to prevent collision with compiler flags.
Previously PyCF_ALLOW_TOP_LEVEL_AWAIT was clashing
with CO_FUTURE_DIVISION.

asottile committed:
bpo-40260: Remove unnecessary newline in compile() call (GH-19641)

Because some people subclass this class and call undocumented methods, and we don't want to break them.

sweeneyde committed:
bpo-39939: Add str.removeprefix and str.removesuffix (GH-18939)

Added str.removeprefix and str.removesuffix methods and corresponding
bytes, bytearray, and collections.UserString methods to remove affixes
from a string if present. See PEP 616 for a full description.

pablogsal committed:
bpo-40334: PEP 617 implementation: New PEG parser for CPython (GH-19503)

Co-authored-by: Guido van Rossum <guido@python.org>
Co-authored-by: Lysandros Nikolaou <lisandrosnik@gmail.com>

pablogsal committed:
PEP 617: Only run the CI with the new parser (GH-19664)

pablogsal committed:
bpo-40334: Fix errors in parse_string.c with old compilers (GH-19666)

pablogsal committed:
Add @pablogsal as code owner for pegen-related files (GH-19665)

pablogsal committed:
bpo-40334: Fix builds outside the source directory and regenerate autoconf files (GH-19667)

vstinner committed:
bpo-40334: Rename PyConfig.use_peg to _use_peg_parser (GH-19670)

* Rename PyConfig.use_peg to _use_peg_parser
* Document PyConfig._use_peg_parser and mark it a deprecated
* Mark -X oldparser option and PYTHONOLDPARSER env var as deprecated
  in the documentation.
* Add use_old_parser() and skip_if_new_parser() to test.support
* Remove sys.flags.use_peg: use_old_parser() uses
  _testinternalcapi.get_configs() instead.
* Enhance test_embed tests
* subprocess._args_from_interpreter_flags() copies -X oldparser

pablogsal committed:
Compile extensions in test_peg_generator with C99 (GH-19668)

pablogsal committed:
bpo-40334: Don't downcast from Py_ssize_t to int (GH-19671)

pablogsal committed:
bpo-40334: Fix build errors and warnings in test_peg_generator (GH-19672)

lysnikolaou committed:
bpo-40334: Suppress all output in test_peg_generator (GH-19675)

The-Compiler committed:
gdbinit: Use proper define syntax (GH-19557)



Using `def` rather than `define` results in:

    Ambiguous command "def pu": define, define-prefix.

Automerge-Triggered-By: @csabella

pablogsal committed:
bpo-40370: Use the same compile and link args as the interpreter used in test_peg_generator (GH-19674)

lysnikolaou committed:
bpo-40334: Improve various PEG-Parser related stuff (GH-19669)

The changes in this commit are all related to @vstinner's original review comments of the initial PEP 617 implementation PR.

corona10 committed:
Update ga_new to use _PyArg_CheckPositional and _PyArg_NoKwnames (GH-19679)

vstinner committed:
bpo-39983: Add test.support.print_warning() (GH-19683)

Log "Warning -- ..." test warnings into sys.__stderr__ rather than
sys.stderr, to ensure to display them even if sys.stderr is captured.

test.libregrtest.utils.print_warning() now calls
test.support.print_warning().

serhiy-storchaka committed:
bpo-40336: Refactor typing._SpecialForm (GH-19620)

gvanrossum committed:
bpo-40334: Use old compiler when compile mode is func_type (GH-19692)

This is invoked by mypy, using ast.parse(source, "<func_type>", "func_type"). Since the new grammar doesn't yet support the func_type_input start symbol we must use the old compiler in this case to prevent a crash.

https://bugs.python.org/issue40334

gvanrossum committed:
Fix broken mkdir -p call in regen-pegen (#19695)

We should use `$(MKDIR_P) <directory>`, not `$(MKDIR_P) -p <directory>`.

pablogsal committed:
bpo-40334: Allow to run make regen-pegen without distutils (GH-19684)

pablogsal committed:
Use Py_ssize_t instead of ssize_t (GH-19685)

vstinner committed:
bpo-40048: Fix _PyCode_InitOpcache() error path (GH-19691)

If _PyCode_InitOpcache() fails in _PyEval_EvalFrameDefault(), use
"goto exit_eval_frame;" rather than "return NULL;" to exit the
function in a consistent state. For example, tstate->frame is now
reset properly.

rhettinger committed:
Expand the implementation comments (GH-19699)

vstinner committed:
bpo-38061: os.closerange() uses closefrom() on FreeBSD (GH-19696)

On FreeBSD, os.closerange(fd_low, fd_high) now calls
closefrom(fd_low) if fd_high is greater than or equal to
sysconf(_SC_OPEN_MAX).

Initial patch by Ed Maste (emaste), Conrad Meyer (cem), Kyle Evans
(kevans) and Kubilay Kocak (koobs):
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=242274

vstinner committed:
bpo-38061: subprocess uses closefrom() on FreeBSD (GH-19697)

Optimize the subprocess module on FreeBSD using closefrom().
A single close(fd) syscall is cheap, but when sysconf(_SC_OPEN_MAX)
is high, the loop calling close(fd) on each file descriptor can take
several milliseconds.

The workaround on FreeBSD to improve performance was to load and
mount the fdescfs kernel module, but this is not enabled by default.

Initial patch by Ed Maste (emaste), Conrad Meyer (cem), Kyle Evans
(kevans) and Kubilay Kocak (koobs):
https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=242274

lysnikolaou committed:
bpo-40334: Rewrite test_c_parser to avoid memory leaks (GH-19694)

Previously every test was building an extension module and
loading it into sys.modules. The tearDown function was thus
not able to clean up correctly, resulting in memory leaks.

With this commit, every test function now builds the extension
module and runs the actual test code in a new process
(using assert_python_ok), so that sys.modules stays intact
and no memory gets leaked.

carljm committed:
bpo-40360: Deprecate lib2to3 module in light of PEP 617 (GH-19663)

Deprecate lib2to3 module in light of PEP 617.

We anticipate removal in the 3.12 timeframe.

caje731 committed:
bpo-40340: Separate examples more clearly in the programming FAQ (GH-19688)

gvanrossum committed:
bpo-40334: Add What's New sections for PEP 617 and PEP 585 (GH-19704)

ammaraskar committed:
closes bpo-40385: Remove Tools/scripts/checkpyc.py (GH-19709)

This is one of the few files that has intimate knowledge of the pyc file
format. Since it lacks tests it tends to become outdated fairly quickly.
At present it has been broken since the introduction of PEP 552.

caje731 committed:
bpo-40279: Add some error-handling to the module initialisation docs example (GH-19705)

serhiy-storchaka committed:
bpo-40275: Avoid importing asyncio in test.support (GH-19600)

* Import asyncio lazily in unittest (only when IsolatedAsyncioTestCase is used).
* Import asyncio.events lazily in test.support.

serhiy-storchaka committed:
bpo-40275: Avoid importing socket in test.support (GH-19603)

* Move socket related functions from test.support to socket_helper.
* Import socket, nntplib and urllib.error lazily in transient_internet().
* Remove importing multiprocess.

serhiy-storchaka committed:
bpo-40275: Avoid importing logging in test.support (GH-19601)

Import logging lazily in assertLogs() in unittest.
Move TestHandler from test.support to logging_helper.

hroskes committed:
Fix typo in object.__format__ docs (GH-19504)

nfishe committed:
Fix typo in Lib/typing.py (GH-19717)

serhiy-storchaka committed:
bpo-40396: Support GenericAlias in the typing functions. (GH-19718)

rhettinger committed:
bpo-40387: Improve queue join() example. (GH-19724)

ammaraskar committed:
bpo-40401: Remove duplicate pyhash.h include from pythoncore.vcxproj (GH-19725)

bsolomon1124 committed:
bpo-38387: Formally document PyDoc_STRVAR and PyDoc_STR macros (GH-16607)

Adds a short description of `PyDoc_STRVAR` and `PyDoc_STR` to "Useful macros" section of C-API docs.

Currently, there is [one lone mention](https://docs.python.org/3/c-api/module.html?highlight=pydoc_strvar#c.PyModuleDef) in the C-API reference, despite the fact that `PyDoc_STRVAR` is ubiquitous to `Modules/`.

Additionally, this properly uses `c:macro` within `Doc/c-api/module.rst` to link.

ZackerySpytz committed:
bpo-40348: Fix typos in the programming FAQ (GH-19729)

serhiy-storchaka committed:
bpo-40398: Fix typing.get_args() for special generic aliases. (GH-19720)

pablogsal committed:
Add files in tests/test_peg_generator to the install target lists (GH-19723)

Update the "Makefile.pre.in" template and the "PCbuild/lib.pyproj" with the files in "Lib/test/test/test_peg_generator" so they get correctly installed along the rest of the standard library.

pablogsal committed:
bpo-40217: Ensure Py_VISIT(Py_TYPE(self)) is always called for PyType_FromSpec types (GH-19414)

pablogsal committed:
bpo-40217: Clean code in PyType_FromSpec_Alloc and add NEWS entry (GH-19733)

corona10 committed:
bpo-40375: Implement imaplib.IMAP4.unselect (GH-19712)

vstinner committed:
bpo-30966: Add multiprocessing.SimpleQueue.close() (GH-19735)



Add a new close() method to multiprocessing.SimpleQueue to explicitly
close the queue.

Automerge-Triggered-By: @pitrou

pablogsal committed:
bpo-40334: Support CO_FUTURE_BARRY_AS_BDFL in the new parser (GH-19721)

This commit also allows to pass flags to the new parser in all interfaces and fixes a bug in the parser generator that was causing to inline rules with actions, making them disappear.

pablogsal committed:
bpo-40334: Support PyPARSE_DONT_IMPLY_DEDENT in the new parser (GH-19736)

vstinner committed:
bpo-30966: concurrent.futures.Process.shutdown() closes queue (GH-19738)

Process.shutdown(wait=True) of concurrent.futures now closes
explicitly the result queue.

vstinner committed:
bpo-39995: Split test_concurrent_futures.test_crash() into sub-tests (GH-19739)

Now only test_error_during_result_unpickle_in_result_handler()
captures and ignores sys.stderr in the test process.

Tools like test.bisect_cmd don't support subTest() but only
work with the granularity of one method.

Remove unused ExecutorDeadlockTest._sleep_id() method.

ambv committed:
Python 3.9.0a6

lysnikolaou committed:
bpo-40334: Catch E_EOF error, when the tokenizer returns ERRORTOKEN (GH-19743)

An E_EOF error was only being caught after the parser exited before this commit. There are some cases though, where the tokenizer returns ERRORTOKEN *and* has set an E_EOF error (like when EOF directly follows a line continuation character) which weren't correctly handled before.

lysnikolaou committed:
bpo-40334: Don't skip test_parser:test_trigget_memory_error (GH-19744)

This test has been changed to always use the old parser, so no need for it to be skipped.

pablogsal committed:
bpo-40334: Refactor peg_generator to receive a Tokens file when building c code (GH-19745)

ambv committed:
Merge tag 'v3.9.0a6'

ambv committed:
Post 3.9.0a6

vstinner committed:
bpo-40421: Add pyframe.h header file (GH-19755)

Add a new separated pyframe.h header file of the PyFrame public C
API: it is included by Python.h.

Add PyFrame_GetLineNumber() to the limited C API.

Replace "struct _frame" with "PyFrameObject" in header files.
PyFrameObject is now defined as struct _frame by pyframe.h which is
included early enough in Python.h.

vstinner committed:
bpo-40421: Add Include/cpython/code.h header file (GH-19756)

bpo-35134, bpo-40421: Add Include/cpython/code.h header file.

code.h now defines PyCodeObject type in the limited C API. It is now
included by Python.h.

Give a name to the PyCodeObject structure: it is now called
"struct PyCodeObject". So it becomes possible to define PyCodeObject
as "struct PyCodeObject" in the limited C API without defining the
structure.

vstinner committed:
bpo-40421: Add PyFrame_GetCode() function (GH-19757)

PyFrame_GetCode(frame): return a borrowed reference to the frame
code.

Replace frame->f_code with PyFrame_GetCode(frame) in most code,
except in frameobject.c, genobject.c and ceval.c.

Also add PyFrame_GetLineNumber() to the limited C API.

abbyonstott committed:
bpo-40025: Require _generate_next_value_ to be defined before members (GH-19098)

require `_generate_next_value_` to be defined before members

tirkarthi committed:
bpo-39966: Revert "bpo-25597: Ensure wraps' return value is used for magic methods in MagicMock" (GH-19734)

* Revert "bpo-25597: Ensure wraps' return value is used for magic methods in MagicMock (#16029)"

This reverts commit 72b1004657e60c900e4cd031b2635b587f4b280e.

vstinner committed:
bpo-40429: PyFrame_GetCode() result cannot be NULL (GH-19772)

Add frame_nslots() to factorize duplicate code.

ericsnowcurrently committed:
bpo-32604: Add support for a "default" arg in channel_recv(). (GH-19770)



This allows the caller to avoid creation of an exception when the channel is empty (just like `dict.get()` works).  `ChannelEmptyError` is still raised if no default is provided.

Automerge-Triggered-By: @ericsnowcurrently

vstinner committed:
bpo-40429: PyFrame_GetCode() now returns a strong reference (GH-19773)

hroncok committed:
bpo-40431: Fix syntax typo in turtledemo (GH-19777)

***   File "/usr/lib64/python3.9/turtledemo/__main__.py", line 275
        bg="#d00" if clear == NORMAL else"#fca")
                                        ^
    SyntaxError: invalid string prefix

LewisGaul committed:
bpo-38880: List interpreters associated with a channel end (GH-17323)



This PR adds the functionality requested by https://github.com/ericsnowcurrently/multi-core-python/issues/52.

Automerge-Triggered-By: @ericsnowcurrently

vstinner committed:
bpo-40429: Refactor super_init() (GH-19776)

Add super_init_without_args() sub-function. Hold a strong reference
to the frame code object while calling super_init_without_args().

vstinner committed:
bpo-40428: Remove PyTuple_ClearFreeList() function (GH-19769)

Remove the following function from the C API:

* PyAsyncGen_ClearFreeLists()
* PyContext_ClearFreeList()
* PyDict_ClearFreeList()
* PyFloat_ClearFreeList()
* PyFrame_ClearFreeList()
* PyList_ClearFreeList()
* PySet_ClearFreeList()
* PyTuple_ClearFreeList()

Make these functions private, move them to the internal C API and
change their return type to void.

Call explicitly PyGC_Collect() to free all free lists.

Note: PySet_ClearFreeList() did nothing.

lysnikolaou committed:
bpo-40334: Fix shifting of nested f-strings in the new parser (GH-19771)

`JoinedStr`s and `FormattedValue also needs to be shifted, in order to correctly compute the location information of nested f-strings.

vstinner committed:
bpo-40429: PyThreadState_GetFrame() returns a strong ref (GH-19781)

The PyThreadState_GetFrame() function now returns a strong reference
to the frame.

pablogsal committed:
bpo-40334: Explicitly cast to int in pegen.c to fix a compiler warning (GH-19779)

shihai1991 committed:
bpo-40275: Move requires_hashdigest() to test.support.hashlib_helper (GH-19716)

Add a new test.support.hashlib_helper submodule.

vstinner committed:
bpo-40421: Add PyFrame_GetBack() function (GH-19765)

New PyFrame_GetBack() function: get the frame next outer frame.

Replace frame->f_back with PyFrame_GetBack(frame) in most code but
frameobject.c, ceval.c and genobject.c.

vstinner committed:
bpo-39995: Fix concurrent.futures _ThreadWakeup (GH-19760)

Fix a race condition in concurrent.futures._ThreadWakeup: access to
_ThreadWakeup is now protected with the shutdown lock.

lysnikolaou committed:
bpo-40334: Disallow invalid single statements in the new parser (GH-19774)

After parsing is done in single statement mode, the tokenizer buffer has to be checked for additional lines and a `SyntaxError` must be raised, in case there are any.

Co-authored-by: Pablo Galindo <Pablogsal@gmail.com>

ZackerySpytz committed:
bpo-40428: Remove references to Py*_ClearFreeList in the docs (GH-19783)

They were removed from the C API in commit
ae00a5a88534fd45939f86c12e038da9fa6f9ed6.

hnarkaytis committed:
Rebase to master

hnarkaytis committed:
Cleanup merge conflict artifacts
